File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.util.Scanner;
import java.util.Stack;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();

        int start = 0;
        Stack<Integer> stack = new Stack<>();
        StringBuilder sb = new StringBuilder();

        while (n-- > 0) {
            int num = sc.nextInt();

            if (num > start) {
                for (int i = start + 1; i <= num; i++) {
                    stack.push(i);
                    sb.append('+').append('\n');
                }
                start = num;
            }

            if (stack.peek() != num) {
                System.out.println("NO");
                return;
            }

            stack.pop();
            sb.append('-').append('\n');
        }
        System.out.print(sb);
    }
}

File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys

# sys.stdin = open("input.txt", "r")

n = int(sys.stdin.readline().strip())

start = 1
stack = []
text = []
isFlag = False

for i in range(n):
    num = int(sys.stdin.readline().strip())
    
    while True:
        if stack and stack[-1] == num:
            stack.pop()
            text.append('-')
            break
        elif stack and stack[-1] > num:
            isFlag = True
            break
        else:
            stack.append(start)
            start += 1
            text.append('+')
    
    if isFlag == True:
        break

if isFlag:
    print('NO')
else:
    for c in text:
        print(c)
